sdk/log/logger_test.go

```diff
@@ -348,3 +357,94 @@ func TestLoggerEnabled(t *testing.T) {
 		})
 	}
 }
+
+func TestLoggerSelfObservability(t *testing.T) {
+	testCases := []struct {
+		name                     string
+		selfObservabilityEnabled bool
+		ctx                      context.Context
+		records                  []log.Record
+		wantLogRecordCount       int64
+	}{
+		{
+			name:                     "SelfObservabilityDisabledZeroLogs",
+			selfObservabilityEnabled: false,
+			ctx:                      context.Background(),
+			records:                  nil,
+			wantLogRecordCount:       0,
+		},
+		{
+			name:                     "SelfObservabilityDisabledNonZeroLogs",
+			selfObservabilityEnabled: false,
+			ctx:                      context.Background(),
+			records:                  []log.Record{{}, {}},
+			wantLogRecordCount:       0,
+		},
+		{
+			name:                     "ZeroLogsCreated",
+			selfObservabilityEnabled: true,
+			ctx:                      context.Background(),
+			records:                  []log.Record{},
+			wantLogRecordCount:       0,
+		},
+		{
+			name:                     "NilContext",
+			selfObservabilityEnabled: true,
+			ctx:                      nil,
+			records:                  []log.Record{{}, {}, {}, {}},
+			wantLogRecordCount:       4,
+		},
+		{
+			name:                     "NonZeroLogsCreated",
+			selfObservabilityEnabled: true,
+			ctx:                      context.Background(),
+			records: []log.Record{
+				{}, {}, {}, {}, {},
+			},
+			wantLogRecordCount: 5,
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(tc.name, func(t *testing.T) {
+			t.Setenv("OTEL_GO_X_SELF_OBSERVABILITY", strconv.FormatBool(tc.selfObservabilityEnabled))
+			prev := otel.GetMeterProvider()
+			defer otel.SetMeterProvider(prev)
+			r := metric.NewManualReader()
+			mp := metric.NewMeterProvider(metric.WithReader(r))
+			otel.SetMeterProvider(mp)
+			l := newLogger(NewLoggerProvider(), instrumentation.Scope{})
+
+			for _, record := range tc.records {
+				l.Emit(tc.ctx, record)
+			}
+
+			gotMetrics := new(metricdata.ResourceMetrics)
+			assert.NoError(t, r.Collect(context.Background(), gotMetrics))
+			if !tc.selfObservabilityEnabled || len(tc.records) == 0 {
+				assert.Empty(t, gotMetrics.ScopeMetrics)
+				return
+			}
+
+			require.Len(t, gotMetrics.ScopeMetrics, 1)
```

Comment:
should it not simply be?

```suggestion
			require.Len(t, gotMetrics.ScopeMetrics, tc.wantLogRecordCount)
```

Reply:
As per my understanding, `r.Collect` reads metrics in gotMetrics which is of type [metricdata.ResourceMetrics](https://pkg.go.dev/go.opentelemetry.io/otel/sdk/metric/metricdata#ResourceMetrics) which basically contains `[]ScopeMetrics` where each element has metrics for unique scope.

In this test, in case if OTEL_GO_X_SELF_OBSERVABILITY is not set, ScopeMetrics slice will be empty else its size will be 1. (for this reason I have check for empty and length 1.
Below these checks, I'm checking for actual metric values and structure.

Let me know if I have misunderstood something here. Happy to fix.

Reply:
I see if `!tc.selfObservabilityEnabled` is even better.